File: JS-Script/hosthunter.js
const dns = require('dns');
const fs = require('fs');
const chalk = require('chalk');
const figlet = require('figlet');
const path = require('path');

const getRandomColor = () => {
    const colors = [
        chalk.red, chalk.green, chalk.yellow, chalk.cyan, chalk.magenta, chalk.blue, chalk.white
    ];
    return colors[Math.floor(Math.random() * colors.length)];
};

figlet('HostName Hunter', function (err, data) {
    if (err) {
        console.log('Something went wrong...');
        console.dir(err);
        return;
    }
    const coloredBanner = getRandomColor()(data);
    console.log(coloredBanner);
});

const getTerminalWidth = () => {
    return process.stdout.columns || 80;
};

const isValidIp = (ip) => {
    const ipRegex = /^(25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)\.(25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)\.(25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)\.(25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)$/;
    return ipRegex.test(ip);
};

const isValidIpRange = (range) => {
    const rangeRegex = /^(25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)\.(25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)\.(25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)\.(25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)(\/([1-9]|[1-2][0-9]|3[0-2]))$/;
    return rangeRegex.test(range);
};

const validateIps = (ips) => {
    return ips.every(ipLine => {
        const ipParts = ipLine.split(',').map(part => part.trim());
        return ipParts.every(ip => {
            const isValid = isValidIp(ip) || isValidIpRange(ip);
            if (!isValid) {
                console.error(chalk.red(`Invalid IP address or range format: "${ip}"`));
            }
            return isValid;
        });
    });
};

const args = process.argv.slice(2);
let ipAddresses = [];
let outputFileName = null;
let results = [['IP Address', 'Status', 'Hostname']];
let completedLookups = 0;

for (let i = 0; i < args.length; i++) {
    if (args[i] === '-i') {
        const input = args[i + 1];
        ipAddresses = input.split(',').map(ip => ip.trim());
        i++;
    } else if (args[i] === '-iF') {
        const filePath = args[i + 1];
        try {
            if (!fs.existsSync(filePath)) {
                console.error(chalk.red(`File not found: ${filePath}`));
                process.exit(1);
            }
            const fileContent = fs.readFileSync(filePath, 'utf-8');
            ipAddresses = fileContent.split(/\r?\n/).map(line => line.trim()).filter(line => line);
            
            const hasCidr = ipAddresses.some(isValidIpRange);
            const hasIps = ipAddresses.some(isValidIp);

            if (hasCidr && hasIps) {
                console.log(chalk.yellow("CIDR is mixed with valid IPs, convert those to IPs."));
                process.exit(1);
            } else if (hasCidr) {
                console.log(chalk.yellow("Please provide the list of IPs, not the CIDR Notation. Can't Handle that as of now."));
                process.exit(1);
            } else if (!validateIps(ipAddresses)) {
                console.error(chalk.red("File contains invalid IP address or range format."));
                process.exit(1);
            }
            i++;
        } catch (error) {
            console.error(chalk.red(`Error reading file: ${error.message}`));
            process.exit(1);
        }
    } else if (args[i] === '-o') {
        outputFileName = args[i + 1] ? args[i + 1] : 'output.csv';
        i++;
    }
}

if (ipAddresses.length === 0) {
    console.error(chalk.red("No IP addresses provided. Use -i <IP-Address> or -iF <File-Path>."));
    process.exit(1);
}

const hasCidrInArgs = ipAddresses.some(isValidIpRange);
if (hasCidrInArgs) {
    console.log(chalk.yellow("I don't handle the CIDR Range."));
    process.exit(1);
}

const handleDnsLookup = (ip) => {
    if (isValidIpRange(ip)) {
        console.log(chalk.yellow(`Can't handle the CIDR range: "${ip}"`));
        results.push([ip, 'Not Applicable', 'CIDR range - Not handled']);
        completedLookups++;
        if (completedLookups === totalLookups) {
            finalizeOutput();
        }
    } else if (isValidIp(ip)) {
        dns.reverse(ip, (err, hostnames) => {
            const status = err ? 'Failed' : 'Success';
            const hostname = err ? `Error: ${err.message}` : hostnames.join(', ');
            results.push([ip, status, hostname]);
            completedLookups++;
            if (completedLookups === totalLookups) {
                finalizeOutput();
            }
        });
    } else {
        console.error(chalk.red(`Invalid IP address or range format encountered: "${ip}"`));
        completedLookups++;
        if (completedLookups === totalLookups) {
            finalizeOutput();
        }
    }
};

let totalLookups = 0;
ipAddresses.forEach(ipLine => {
    const ipParts = ipLine.split(',').map(part => part.trim());
    totalLookups += ipParts.length;
    ipParts.forEach(ip => handleDnsLookup(ip));
});

function finalizeOutput() {
    printTable();
    if (outputFileName) saveToCSV();
}

function printTable() {
    if (results.length <= 1) {
        console.log(chalk.yellow("No results to display."));
        return;
    }

    const maxIpLength = Math.max(...results.map(row => row[0].length), 'IP Address'.length);
    const maxStatusLength = Math.max(...results.map(row => row[1].length), 'Status'.length);
    const maxHostnameLength = Math.max(...results.map(row => row[2].length), 'Hostname'.length);

    const rowSeparator = `+${'-'.repeat(maxIpLength + 2)}+${'-'.repeat(maxStatusLength + 2)}+${'-'.repeat(maxHostnameLength + 2)}+`;

    console.log(rowSeparator);
    console.log(`| ${'IP Address'.padEnd(maxIpLength)} | ${'Status'.padEnd(maxStatusLength)} | ${'Hostname'.padEnd(maxHostnameLength)} |`);
    console.log(rowSeparator);

    results.slice(1).forEach(row => {
        const colorStatus = row[1] === 'Success' 
            ? chalk.green(row[1].padEnd(maxStatusLength)) 
            : chalk.red(row[1].padEnd(maxStatusLength));

        console.log(`| ${row[0].padEnd(maxIpLength)} | ${colorStatus} | ${row[2].padEnd(maxHostnameLength)} |`);
        console.log(rowSeparator);
    });
}

function saveToCSV() {
    const csvData = results.map(row => row.join(',')).join('\n');
    const filePath = path.resolve(__dirname, outputFileName);
    fs.writeFileSync(filePath, csvData);
    console.log(chalk.green(`Results saved to ${filePath}`));
}

File: Python-Script/hosthunter.py
import socket
import sys
import os
import csv
import pyfiglet
import random
from colorama import Fore, Style, init

# Initialize colorama for cross-platform colored output
init(autoreset=True)

def get_random_color():
    colors = [Fore.RED, Fore.GREEN, Fore.YELLOW, Fore.CYAN, Fore.MAGENTA, Fore.BLUE, Fore.WHITE]
    return random.choice(colors)

# Generate dynamic ASCII banner using pyfiglet
ascii_banner = pyfiglet.figlet_format("HostName Hunter")
colored_banner = get_random_color() + ascii_banner + Style.RESET_ALL

# Display the tool name banner
print(colored_banner)

def is_valid_ip(ip):
    try:
        socket.inet_aton(ip)
        return True
    except socket.error:
        return False

def is_valid_ip_range(ip):
    return '/' in ip

def perform_dns_lookup(ip):
    try:
        hostname = socket.gethostbyaddr(ip)[0]
        return "Success", hostname
    except socket.herror:
        return "Failed", "No hostname found"

def load_ips_from_file(file_path):
    with open(file_path, 'r') as file:
        return [line.strip() for line in file if line.strip()]

def print_table(results):
    max_ip_len = max(len(row[0]) for row in results)
    max_status_len = max(len(row[1]) for row in results)
    max_hostname_len = max(len(row[2]) for row in results)

    row_separator = f"+{'-' * (max_ip_len + 2)}+{'-' * (max_status_len + 2)}+{'-' * (max_hostname_len + 2)}+"
    print(row_separator)
    print(f"| {'IP Address'.ljust(max_ip_len)} | {'Status'.ljust(max_status_len)} | {'Hostname'.ljust(max_hostname_len)} |")
    print(row_separator)

    for row in results:
        ip, status, hostname = row
        color_status = Fore.GREEN + status if status == "Success" else Fore.RED + status
        print(f"| {ip.ljust(max_ip_len)} | {color_status.ljust(max_status_len + len(Fore.GREEN))} | {hostname.ljust(max_hostname_len)} |")
        print(row_separator)

def save_to_csv(results, output_file):
    with open(output_file, 'w', newline='') as csvfile:
        writer = csv.writer(csvfile)
        writer.writerow(['IP Address', 'Status', 'Hostname'])
        writer.writerows(results)
    print(Fore.GREEN + f"Results saved to {output_file}")

def main():
    if len(sys.argv) < 3:
        print(Fore.RED + "Usage: python hosthunter.py -i <IP-Address or IP-Addresses> or -iF <File-Path> [-o <Output-File>]")
        sys.exit(1)

    args = sys.argv[1:]
    ip_addresses = []
    output_file = None

    if '-i' in args:
        index = args.index('-i') + 1
        if index >= len(args):
            print(Fore.RED + "No IP address provided after -i.")
            sys.exit(1)
        ip_addresses = [ip.strip() for ip in args[index].split(',')]
    
    elif '-iF' in args:
        index = args.index('-iF') + 1
        if index >= len(args):
            print(Fore.RED + "No file path provided after -iF.")
            sys.exit(1)
        file_path = args[index]
        if not os.path.exists(file_path):
            print(Fore.RED + f"File not found: {file_path}")
            sys.exit(1)
        ip_addresses = load_ips_from_file(file_path)
    
    if not ip_addresses:
        print(Fore.RED + "No IP addresses provided.")
        sys.exit(1)

    if '-o' in args:
        index = args.index('-o') + 1
        if index < len(args):
            output_file = args[index]
        else:
            output_file = 'output.csv'

    # Check for CIDR and mixed IP/CIDR content
    has_cidr = any(is_valid_ip_range(ip) for ip in ip_addresses)
    has_ips = any(is_valid_ip(ip) for ip in ip_addresses)

    if has_cidr and has_ips:
        print(Fore.YELLOW + "Only IPs are accepted, convert the CIDR into the list of IPs and update the file.")
        sys.exit(1)
    elif has_cidr:
        print(Fore.YELLOW + "I don't handle the CIDR Range.")
        sys.exit(1)

    results = [['IP Address', 'Status', 'Hostname']]
    
    for ip in ip_addresses:
        if is_valid_ip(ip):
            status, hostname = perform_dns_lookup(ip)
            results.append([ip, status, hostname])
        else:
            print(Fore.RED + f"Invalid IP address format: {ip}")
            results.append([ip, "Invalid", "N/A"])

    print_table(results[1:])

    if output_file:
        save_to_csv(results[1:], output_file)

if __name__ == "__main__":
    main()
